### PR TITLE
PR For Private-Event project review

### PR DESCRIPTION
**Pull request by umairarshadbutt**
The purpose of PR is to evaluate the practice project (Private Event).
This project is developed for a good understanding of Active record Association

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,75 @@
-# Private-Events
\ No newline at end of file
+<h3 align="center">Private Events - RoR</h3>
+
```
Comment (christianotieno) [inline README.md:2]: ## Add a screenshot of your page on the README 

- [ ] Add a screenshot of your page here. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_readme.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/)).

File: README.md [hunk 2]
```diff
@@ -1 +1,75 @@
...
+### Setup
+
+- Open Terminal
+- Change directory to folder to download repository
+- Use `cd <file-path>`
+- Run `https://github.com/umairarshadbutt/Private-Events.git`
+- cd to the project `cd Private-Events`
+- Install gems `bundle install`
+- Install Yarn `yarn install --check-files` 
+- Run `rails server`
```
Comment (christianotieno) [inline README.md:39]: ## Add test instructions
- [x] Once implemented, be sure to also add instructions to run the tests as well. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_readme.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/)).

File: app/views/events/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    </tr>
+  </thead>
+  <tbody>
+    <tr><% @events.each do |e| %></tr>
+      <td><%= e.title %></td>
+      <td><%= e.description %></td>
+      <td><%= e.event_date %></td>
+      <td><%= e.creator.username %></td>
+      <td><%= link_to "Show details", event_path(e) %></td>
+    <% end %>
```
Comment (christianotieno) [inline app/views/events/index.html.erb]: ## Remove all the logic related code 

* [x] View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)
* [x] There shoud be absolutely NO logic in views.

  > Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* Check this article on [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code) and [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) for more info about this. Thank you. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright.

File: app/views/events/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      </tr>
+    </thead>
+    <tbody>
+      <tr><% @all_upcoming_events.each do |e| %></tr>
+        <td><%= e.title %></td>
+        <td><%= e.description %></td>
+        <td><%= e.event_date %></td>
+        <td><%= e.creator.username %></td>
+        <td><%= link_to "Show details", event_path(e) %></td>
+      <% end %>
```
Comment (christianotieno) [inline app/views/events/index.html.erb]: ## Remove all the logic related code 

* [x] View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)
* [x] There shoud be absolutely NO logic in views.

  > Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* Check this article on [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code) and [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) for more info about this. Thank you. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright.

File: app/views/events/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      </tr>
+    </thead>
+    <tbody>
+      <tr><% @all_past_events.each do |e| %></tr>
+        <td><%= e.title %></td>
+        <td><%= e.description %></td>
+        <td><%= e.event_date %></td>
+        <td><%= e.creator.username %></td>
+        <td><%= link_to "Show details", event_path(e) %></td>
+      <% end %>
```
Comment (christianotieno) [inline app/views/events/index.html.erb]: ## Remove all the logic related code 

* [x] View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)
* [x] There shoud be absolutely NO logic in views.

  > Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* Check this article on [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code) and [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) for more info about this. Thank you. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright.

File: app/views/events/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+<% if @event.attendees.empty? %>
+  <p>No attandees</p>
+<% else %>
+  <strong>Attendees:</strong>
+  <ul>
+    <% @event.attendees.each do |attendee| %>
+      <li><%= link_to attendee.full_name, user_path(attendee.id) %></li>
+    <% end %>
+  </ul>
```
Comment (christianotieno) [inline app/views/events/show.html.erb]: ## Remove all the logic related code 

* [x] View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)
* [x] There shoud be absolutely NO logic in views.

  > Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* Check this article on [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code) and [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) for more info about this. Thank you. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright.

File: app/views/events/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+<% end %>
```
Comment (christianotieno) [inline app/views/events/show.html.erb]: ## Remove all the logic related code 

* [x] View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)
* [x] There shoud be absolutely NO logic in views.

  > Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* Check this article on [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code) and [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) for more info about this. Thank you. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright.

File: test/application_system_test_case.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+require 'test_helper'
```
Comment (christianotieno) [inline test/application_system_test_case.rb:1]: #### Tests

* [x] Make sure models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [x] Make sure authentication have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [x] Make sure event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

**Review by christianotieno — CHANGES_REQUESTED** [review]
# Changes Required ♻️ 
Hello team. 👋 
The word done is really solid. Keep it up. 👏 
Before we merge this, let's work on the highlighted issues below 👇 thank you.

### Happy coding 💻 

#### Christian Otieno (TSE)

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Request changes :recycle:
### Good job so far :+1: However, before this project can be approved you need to consider the required changes outlined in the previous review.

- **_Remove all the logic from the views, use helpers instead_**

- **_Fix the tests; it output errors when I ran the tests._**

#### _Reviewed by Temitayo Adeleke._

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team    
#### Great Job so far, everything works fine, but there are a few more things required. 
### General 
- [ ] Let's work on all requirements as requested by previous TSE's. 
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team    
#### Great Job so far, everything works fine, but there are a few more things required. 
### General 
Let's work on all requirements as requested by previous TSE's. 
- [ ] The test suite is still failing 
- [ ] There should be NO application logic in the views nor helpers
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by umairarshadbutt** [comment]
I'm using a loop for displaying events in views. I think we can use the loop in reviews.

File: app/views/users/_show.html.erb [hunk 9]
```diff
@@ -0,0 +1,56 @@
+<% if current_user? %>
+  <h3 class="text-center">User profile page for: <%= @user.full_name if @user %></h3>
```
Comment (kubilaycaglayan) [inline app/views/users/_show.html.erb]: :bulb: This is a "logic in the views" situation.

Comment (kubilaycaglayan) [reply app/views/users/_show.html.erb]: Both ifs.

File: app/views/users/_new.html.erb [hunk 10]
```diff
@@ -0,0 +1,35 @@
+<% if current_user == nil && !current_user? %>
+  <nav class="navbar navbar-expand-lg navbar-light bg-light">
```
Comment (kubilaycaglayan) [inline app/views/users/_new.html.erb]: :bulb:  This is a "logic in the views" situation.

File: app/views/sessions/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,26 @@
+<% if current_user == nil && !current_user? %>
```
Comment (kubilaycaglayan) [inline app/views/sessions/new.html.erb]: :bulb:  This is a "logic in the views" situation.

File: app/views/sessions/welcome.html.erb [hunk 12]
```diff
@@ -0,0 +1,30 @@
+<% if current_user == nil && !current_user? %>
```
Comment (kubilaycaglayan) [inline app/views/sessions/welcome.html.erb]: :bulb:  This is a "logic in the views" situation.

File: app/views/events/_show.html.erb [hunk 13]
```diff
@@ -0,0 +1,18 @@
+<% if current_user? %>
```
Comment (kubilaycaglayan) [inline app/views/events/_show.html.erb]: :bulb: This is a "logic in the views" situation.

File: app/views/events/_new.html.erb [hunk 14]
```diff
@@ -0,0 +1,22 @@
+
+<% if current_user? %>
```
Comment (kubilaycaglayan) [inline app/views/events/_new.html.erb]: :bulb: This is a "logic in the views" situation.

File: app/views/events/_index.html.erb [hunk 15]
```diff
@@ -0,0 +1,69 @@
+<% if current_user? %>
```
Comment (kubilaycaglayan) [inline app/views/events/_index.html.erb]: :bulb: This is a "logic in the views" situation.

**Review by kubilaycaglayan — CHANGES_REQUESTED** [review]
Hello Team :wave: 
Status: Changes Required :recycle: 

Thank you very much for all your effort! 
I want to point out 2 things, let's talk about them ⏬

### No Logic in The Views

- [x]  There should be absolutely NO logic in views.

Just like you are doing it very well by creating partials, you should create the logics in the helpers to separate and organize the code 🙏

:bulb: Please check the inline comments below, I tried to point out some of the `if` statements that I saw in the views. Please move all of them as functional methods to the helpers. 

:point_right: You can use this source to learn how: https://www.rubyguides.com/2020/01/rails-helpers/
Also:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
- [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

### Tests 
There are two things about tests:

- [ ]  You have a folder named `test`. I believe this is coming from the rails template. And it is distorting the `rspec` command. Please delete it if you are not using it. I had to delete it to run the instructions in the readme file in a healthy way.

- [x]  One test case is failing on my side. I will share a screenshot to give you more insight into it.
**Screenshot from test results :camera_flash:**
image

### That's it, team! After making these changes, please request another code review :angel: :blush: 

With all my respect :pray:

**Comment by umairarshadbutt** [comment]
- Move these views logic into helper with the help of [StackOverflow Solution](https://stackoverflow.com/questions/34597553/how-should-i-remove-this-logic-from-my-view-in-rails)
- Remove this one test case  :camera_flash:

image

File: app/views/events/_index.html.erb [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+      </tr>
+    </thead>
+    <tbody>
+      <tr><% @events.each do |e| %></tr>
+        <td><%= e.title %></td>
+        <td><%= e.description %></td>
+        <td><%= e.event_date %></td>
+        <td><%= e.creator.username %></td>
+        <td><%= link_to "Show details", event_path(e) %></td>
+      <% end %>
```
Comment (BrittanyBlake) [inline app/views/events/_index.html.erb:21]: - [ ] There should be no logic in the views. Please encapsulate all of the view-related logic in your view helpers: **(app/helpers/*_helper.rb)**

File: app/views/events/_index.html.erb [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+      </thead>
+      <tbody>
+        <tr><% @all_upcoming_events.each do |e| %></tr>
+          <td><%= e.title %></td>
+          <td><%= e.description %></td>
+          <td><%= e.event_date %></td>
+          <td><%= e.creator.username %></td>
+          <td><%= link_to "Show details", event_path(e) %></td>
+        <% end %>
+      </tbody>
```
Comment (BrittanyBlake) [inline app/views/events/_index.html.erb:45]: - [ ] There should be no logic in the views. Please encapsulate all of the view-related logic in your view helpers: **(app/helpers/*_helper.rb)**

File: app/views/events/_index.html.erb [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        </tr>
+      </thead>
+      <tbody>
+        <tr><% @all_past_events.each do |e| %></tr>
+          <td><%= e.title %></td>
+          <td><%= e.description %></td>
+          <td><%= e.event_date %></td>
+          <td><%= e.creator.username %></td>
+          <td><%= link_to "Show details", event_path(e) %></td>
+        <% end %>
```
Comment (BrittanyBlake) [inline app/views/events/_index.html.erb:65]: - [ ] There should be no logic in the views. Please encapsulate all of the view-related logic in your view helpers: **(app/helpers/*_helper.rb)**

File: app/views/events/_show.html.erb [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+    <p>No attandees</p>
+  <% else %>
+    <strong>Attendees:</strong>
+    <ul>
+      <% @event.attendees.each do |attendee| %>
+        <li><%= link_to attendee.full_name, user_path(attendee.id) %></li>
+      <% end %>
+    </ul>
+  <% end %>
+
```
Comment (BrittanyBlake) [inline app/views/events/_show.html.erb]: - [ ] There should be no logic in the views. Please encapsulate all of the view-related logic in your view helpers: **(app/helpers/*_helper.rb)**

File: app/views/sessions/welcome.html.erb [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+      </ul>
+      <span class="form-inline my-2 my-lg-0">
+        <span class="btn btn-outline-success my-2 my-sm-0 mr-2"><%= link_to 'Log In', login_path %> </span>
+        <span class="btn btn-outline-success my-2 my-sm-0"><%= link_to 'Sign Up', new_user_path %></span> 
+      </span>
+    </div>
+  </nav>
+<% end %>
+<h1 class="text-center">Welcome to Private Event App</h1>
+  <% if current_user %>
```
Comment (BrittanyBlake) [inline app/views/sessions/welcome.html.erb]: - [ ] There should be no logic in the views. Please encapsulate all of the view-related logic in your view helpers: **(app/helpers/*_helper.rb)**

File: app/views/users/_show.html.erb [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+          <th>Description</th>
+          <th>Event Date</th>
+        </tr>
+      </thead>
+      <tbody>
+          <tr><% @created_events.each do |e| %></tr>
+            <td><%= e.title %></td>
+            <td><%= e.description %></td>
+            <td><%= e.event_date %></td>
+            <% end %>
```
Comment (BrittanyBlake) [inline app/views/users/_show.html.erb:18]: - [ ] There should be no logic in the views. Please encapsulate all of the view-related logic in your view helpers: **(app/helpers/*_helper.rb)**

File: app/views/users/_show.html.erb [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+            <th>Description</th>
+            <th>Event Date</th>
+          </tr>
+        </thead>
+        <tbody>
+            <tr><% @upcoming_events.each do |e| %></tr>
+              <td><%= e.title %></td>
+              <td><%= e.description %></td>
+              <td><%= e.event_date %></td>
+            <% end %>
```
Comment (BrittanyBlake) [inline app/views/users/_show.html.erb:35]: - [ ] There should be no logic in the views. Please encapsulate all of the view-related logic in your view helpers: **(app/helpers/*_helper.rb)**

File: app/views/users/_show.html.erb [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+            <th>Place</th>
+            <th>Event Date</th>
+          </tr>
+        </thead>
+        <tbody>
+            <tr><% @previous_events.each do |e| %></tr>
+              <td><%= e.title %></td>
+              <td><%= e.description %></td>
+              <td><%= e.event_date %></td>
+            <% end %>
```
Comment (BrittanyBlake) [inline app/views/users/_show.html.erb:52]: - [ ] There should be no logic in the views. Please encapsulate all of the view-related logic in your view helpers: **(app/helpers/*_helper.rb)**

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
### Status: Changes Required ♻️ 
****
### Hi Team!👋 
You have done a great job implementing the changes requested by the previous TSE! 👍 You have also done a wonderful job with your project so far! It looks really good! 👏 👏 👏  There are just a few issues that need to be resolved before you can move on: ⬇️ 
****

-  As was pointed out to you in previous reviews, there should be no logic in the views. Take a look at my comments for instances where the logic needs to be removed. 

****
### Please make the required changes and resubmit for a new code review once you are ready 🙏

Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

**Comment by umairarshadbutt** [comment]
I moved the logic from views to helpers. Now there are loops in views

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello  Umair Arshad** 
Great job following the improvements of the last review :muscle:  :muscle: 
Your project meets most of the requirements :clap: :clap: :clap:  
We still need to fix some details:

- [ ] user show route should also show the past and upcoming attendees from the user.
- [ ] Add unit tests for methods that retrieve past and upcoming events and attendees.

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Comment by umairarshadbutt** [comment]
Hi
I'm sharing my system the user route screenshot
user_events

**Review by MuriloRoque — APPROVED** [review]
# Status: Approved :green_circle: 

### Hello @umairarshadbutt :wave: 

### Thank you for making all of the required changes, your project is approved :clap: :clap: :clap: 

### You may merge, good luck with your next projects :rocket: 

Murilo Roque Paiva da Silva

- Github: [@muriloroque](https://github.com/MuriloRoque)
- Linkedin: [MuriloRoque](https://www.linkedin.com/in/murilo-roque-b1268741/)

[Twitter Follow](https://twitter.com/MuriloRoquePai1)

**Pull request merged**